evennia/contrib/auditing/server.py

```diff
@@ -0,0 +1,266 @@
+        respective hook methods should be connected to the methods unique
+        for the respective protocol so that there is a unified interface
+        to Evennia.
+     2. A Server session. This is the same for all connected accounts,
+        regardless of how they connect.
+
+    The Portal and Server have their own respective sessionhandlers. These
+    are synced whenever new connections happen or the Server restarts etc,
+    which means much of the same information must be stored in both places
+    e.g. the portal can re-sync with the server when the server reboots.
+
+    This particular implementation parses all server inputs and/or outputs and 
+    passes a dict containing the parsed metadata to a callback method of your 
+    creation. This is useful for recording player activity where necessary for 
+    security auditing, usage analysis or post-incident forensic discovery.
+    
+    *** WARNING ***
+    All strings are recorded and stored in plaintext. This includes those strings
+    which might contain sensitive data (create, connect, @password). These commands
+    have their arguments masked by default, but you must mask or mask any
+    custom commands of your own that handle sensitive information.
+    
+    Installation:
+
+    Designate this class as the SERVER_SESSION_CLASS in `settings.py`, then set
+    some additional options concerning what to log and where to send it.
+    
+    settings.py:
+    SERVER_SESSION_CLASS = 'evennia.contrib.auditing.server.AuditedServerSession'
+    
+    # Where to send logs? Define the path to a module containing a function 
+    # called 'output()' you've written that accepts a dict object as its sole 
+    # argument. From that function you can store/forward the message received
+    # as you please. An example file-logger is below:
+    AUDIT_CALLBACK = 'evennia.contrib.auditing.examples'
+    
+    # Log all user input? Be ethical about this; it will log all private and 
+    # public communications between players and/or admins.
+    AUDIT_IN = True/False
+    
+    # Log all server output? This will result in logging of ALL system
+    # messages and ALL broadcasts to connected players, so on a busy MUD this 
+    # will be very voluminous!
+    AUDIT_OUT = True/False
+    
+    # What commands do you NOT want masked for sensitivity?
+    AUDIT_MASK_IGNORE = ['@ccreate', '@create']
+    
+    # What commands do you want to keep the first two terms of, masking the rest?
+    # This only triggers if there are more than two terms in the message.
+    AUDIT_MASK_KEEP_BIGRAM = ['create', 'connect', '@userpassword']
+    """
+    def audit(self, **kwargs):
+        """
+        Extracts messages and system data from a Session object upon message 
+        send or receive.
+    
+        Kwargs:
+            src (str): Source of data; 'client' or 'server'. Indicates direction.
+            text (list): Message sent from client to server.
```

Comment:
Write this as `text (str or list): ... ` and then detail that the type depends on where the data comes from. 

It's also very important to note in the doc that this implementation only handles the `text` input/outputfunc but that **kwargs can well include other types of data.

Reply:
How do I document arbitrary pass-through kwargs? This function only concerns itself with `src` and `text`, but I do want to allow anything else provided by a dev to flow through to the logger (since `text` may not be what they want to capture.)